### ISSUE TITLE
more flexible validation options for open text inputs

### ISSUE THREAD

**Issue by alexjsdev** [lookfirst/mui-rff#186]
**Is your feature request related to a problem? Please describe.**
First of all, great work on this lib, it's pretty useful and i'm glad i found it instead of setting this stuff up myself within my app. Thanks for all of your work on it!

Unfortunately, I cannot update to 2.0 because of the way validation has changed in the 2.* versions. It's a very simple and minor difference, but it's quite impactful for the end user experience in mine (and probably most apps).

[Here:](https://github.com/lookfirst/mui-rff/blob/master/src/Util.tsx#L41)
```
export function showError({ meta: { submitError, dirtySinceLastSubmit, error, touched, modified } }: showErrorProps) {
	return !!(((submitError && !dirtySinceLastSubmit) || error) && (touched || modified));
}
```

Just by adding that `|| modified` makes it so that field validation is triggered as soon as the user starts typing, which in the case of validation like "email" for instance, the user doesn't even get one chance (until they focus out of the field) to get it right. Which results in overly aggressive validation. As I type this I'm wondering if the intention was maybe to check if it was `touched && modified` instead of the OR?

**Describe the solution you'd like**
I think it would be sweet if we could either drop the `modified` check, or make that part optional either via prop, or an override that could be provided to determine when a field is in an error state.

I'm likely going to modify what i need in a fork and publish it to our private NPM at work, but would be interested in your thoughts, or if there's a simple workaround i'm missing. Thanks!

Edit: btw i would love to help and submit a PR for this, just want to make sure i'm not missing anything first. Let me know.

**Comment by lookfirst** [comment]
Hi @alexjsdev! Thanks for the report!

I'd hate for you to have to maintain your own fork. I'm totally open to fixing this and making things right.

The big issue is that validation is a touchy/modified subject (punny!)... what if you do want to validate on modified? Maybe `&&` is the right solution there ... need to test and see what works right.

At the end of the day, what I think really needs to happen is that this showError function needs to get abstracted out such that it can be defined externally. 2.x got us that function extracted, so that was step 1. Step 2 is to make it so that people can override the behavior.

I think eventually I want to get this library to a state where something can be written on top of it to build forms with a DSL. I recently learned about [Data Driven Forms](https://github.com/data-driven-forms/react-forms/), which is effectively a duplicate of this project, with even more functionality on top of it. I'm not sure I fully agree with their implementation (no typescript and it is a bit too complex imho), but having a DSL to describe forms is a great idea. Getting that DSL to do validation in all the various cases people can come up with, is going to be tough though.

**Comment by alexjsdev** [comment]
Hey @lookfirst thanks for getting back to me! 😆that is punny!  haha 

Although I see what you mean, to me it seems that the validation is just too aggressive for text inputs specifically and I would think most users wouldn't want validation yelling at them while they haven't even had a chance to finish typing, (so the app doesn't even know yet if they are going to end up with a correct email, for instance, or not as they attempt to move on to the next field in the form). So if we're going to be opinionated here might as well take the majority side, maybe I'm wrong but this just doesn't make sense to me: 

aggressive validation (current behavior): 

![mui-rff-aggro-validation](https://user-images.githubusercontent.com/4146729/79492399-f1448e00-7fed-11ea-8c37-8522b05646b2.gif)

this is just changing the `||` to an `&&` (but i found a bug with this, at least for my implementation) 

![mui-rff-nicer-validation](https://user-images.githubusercontent.com/4146729/79495273-5bf7c880-7ff2-11ea-896b-ee3eb30980fd.gif)

As you can see the bug is that now i _have_ to touch AND modify the field for the validation message to be triggered and render. You can see in my last use case i just hit next without interacting with the already erroneous input. So a simple solution would be to add a prop that could be passed to decide if the validation trigger is `touched` or `modified` (this would accommodate how it worked pre v2 and current behavior). 

Generally I totally agree with you on this: :100: and this is what i was wondering if i was missing somehow:
> At the end of the day, what I think really needs to happen is that this showError function needs to get abstracted out such that it can be defined externally. 2.x got us that function extracted, so that was step 1. Step 2 is to make it so that people can override the behavior.

Something else to consider is that for more complex forms (like the app I work on) you might have a scenario where parts of one larger form (think final form state) might be rendered one at a time like in the case of a wizard form, where there is let's say 1 question per page. I  don't set up validation for all my fields at the top level of the form, I set validation on each field based on a map of rules for each one as it comes from the backend, and i never render the entire form at once. So I pass a validators prop (which is a function in the case of react final form) down to each field component. Although the DSL approach is cool personally I prefer declarative components over configuration.

TL;DR: i think it would be nice if short term, we could add a prop that can be passed to any field that can toggle between including `modified` in the calculation at all or not, maybe something like `validationType="live"` (current behavior) or `validationType="lazy"` (how it used to work before v2).  And longer term include a way to override the calculation of error state and message, and be able to pass that in without wrapping the mui-rff field in another field of the same name, to get access to the meta properties needed from RFF. 

Let me know if anything above doesn't make sense. Hopefully we can agree on one of the options i described above, and i'd be happy to work on a PR to implement it, let me know which direction you'd like to go. Thanks!

**Comment by alexjsdev** [comment]
FYI i've opened a PR here to quickly address this: https://github.com/lookfirst/mui-rff/pull/191

**Comment by mickrip** [comment]
I came across the same problem/gripe regarding 'premature' validation errors, and it led me here -
@alexjsdev is right re: `|| modified`.

I also agree with @lookfirst regarding abstracting this functionality out and giving the user some sort of control over when to show the validation error.

I think one thing we can agree on (hopefully) is that validation shouldn't appear the moment a user starts typing. As a user, it's distracting and negative. It should validate once we suspect beyond a reasonable doubt that the user is finished with that input. I think the default behaviour should reflect this. However, that being said, it would be nice to give it some control.

Why not instead of `validationType="live"`, instead of a string, we use the function itself? (and rename the prop to validationTrigger?)

**Comment by alexjsdev** [comment]
yes @mickrip generally i agree. Though i can see valid use case for instant validation feedback as well, like you pointed out the user (dev) should have control of when to show validation.  I updated the approach to exactly what you described here: https://github.com/lookfirst/mui-rff/pull/235 just waiting for any feedback, or for a merge.

**Comment by lookfirst** [comment]
Closing, this hasn't come up in years. If it is still an issue, we can reopen.
